tests/unit/tuf_repository_service_worker/test_repository.py

```diff
@@ -3428,31 +3412,312 @@ def fake_get_fresh(key):
+            lambda *a: "fake"
+        )
+
+        test_repo._root_metadata_update = pretend.call_recorder(
+            lambda a: {
+                "message": "Metadata Update Processed",
+                "role": "root",
+            }
+        )
+
+        # Use `next` below to mock subsequent calls
+        fake_signature_result = iter((True, False))
+        fake_threshold_result = iter((False, False))
+
+        test_repo._validate_signature = pretend.call_recorder(
+            lambda *a: next(fake_signature_result)
+        )
+        test_repo._validate_threshold = pretend.call_recorder(
+            lambda *a: next(fake_threshold_result)
+        )
+
+        # Call sign_metadata with fake payload
+        # All deserialization and validation is mocked
+        payload = {"signature": "fake", "role": "root"}
+        result = test_repo.sign_metadata(payload)
+
+        assert result == {
+            "task": "sign_metadata",
+            "status": True,
+            "last_update": mocked_datetime.now(),
+            "details": {
+                "message": "Signature Processed",
+                "update": "Root v2 is pending signatures",
+            },
+        }
+
+    def test_sign_metadata__update__invalid_threshold__trusted(
+        self,
+        test_repo,
+        monkeypatch,
+        mocked_datetime,
+    ):
+        def fake_get_fresh(key):
+            if key == "BOOTSTRAP":
+                return "<task-id>"
+            if key == "ROOT_SIGNING":
+                return {"metadata": "fake"}
+
+        fake_settings = pretend.stub(
+            get_fresh=pretend.call_recorder(fake_get_fresh),
+        )
+        monkeypatch.setattr(
+            repository,
+            "get_repository_settings",
+            lambda *a, **kw: fake_settings,
+        )
+        fake_signature = pretend.stub(keyid="fake")
+        repository.Signature.from_dict = pretend.call_recorder(
+            lambda *a: fake_signature
+        )
+        fake_trusted_root = repository.Metadata(
+            signed=repository.Root(version=1)
+        )
+        test_repo._storage_backend.get = pretend.call_recorder(
+            lambda r: fake_trusted_root
+        )
+        fake_new_root = repository.Metadata(signed=repository.Root(version=2))
+        repository.Metadata.from_dict = pretend.call_recorder(
+            lambda *a: fake_new_root
+        )
+        fake_new_root.to_dict = pretend.call_recorder(lambda: "fake")
+        test_repo.write_repository_settings = pretend.call_recorder(
+            lambda *a: "fake"
+        )
+
+        test_repo._root_metadata_update = pretend.call_recorder(
+            lambda a: {
+                "message": "Metadata Update Processed",
+                "role": "root",
+            }
+        )
+
+        # Use `next` below to mock subsequent calls
+        fake_signature_result = iter((False, True))
+        fake_threshold_result = iter((False, True))
```

Comment:
Interesting way of mocking return type.
I like it better than compared to using `MagicMock` as we did until now when we have to mock a different behavior of two calls.